File: frontend/src/api/deployment-queue.ts
/**
 * Deployment Queue API endpoints.
 *
 * Following DEVELOPERS.MD principles:
 * - Type hints everywhere
 * - Explicit operations
 */

import { api } from "../lib/api-client";
import type {
  CsvImportPreview,
  CsvImportResult,
  DeploymentImportRow,
} from "./types";

export type { DeploymentImportRow };

export interface DeploymentQueueEntry {
  id: string;
  project_id: string;
  folder_path: string;
  video_count: number;
  image_count: number;
  site_id: string | null;
  datetime_offset_seconds: number | null;
  /** Fill missing capture dates from each file's modification time.
   * Never overrides a real capture date. */
  use_file_mtime_fallback: boolean;
  /** The subfolders are dependent cameras. Copied onto the
   * deployment when the entry is processed. */
  paired_cameras: boolean;
  /** Per-camera seconds on top of datetime_offset_seconds. */
  camera_offsets: Record<string, number>;
  notes: string | null;
  tags: Record<string, string>;
  status: "pending" | "processing" | "completed" | "failed";
  created_at_utc: string;
  processed_at_utc: string | null;
  error: string | null;
  /** Newline-joined paths of files skipped during ingest for non-fatal
   * reasons (e.g. no extractable capture timestamp). Null when nothing
   * was skipped. */
  warnings: string | null;
  deployment_id: string | null;
}

export interface DeploymentQueueCreate {
  project_id: string;
  folder_path: string;
  site_id?: string | null;
  video_count?: number;
  image_count?: number;
  datetime_offset_seconds?: number | null;
  use_file_mtime_fallback?: boolean;
  paired_cameras?: boolean;
  camera_offsets?: Record<string, number>;
  notes?: string | null;
  tags?: Record<string, string>;
}

export interface ProcessQueueRequest {
  project_id: string;
}

export interface ProcessQueueResponse {
  message: string;
  jobs_started: number;
  job_ids: string[];
  queue_entry_ids: string[];
}

export const deploymentQueueApi = {
  /**
   * List all queue entries for a project
   */
  list: (projectId: string, status?: string) => {
    const params = new URLSearchParams({ project_id: projectId });
    if (status) params.append("status", status);
    return api.get<DeploymentQueueEntry[]>(
      `/api/deployment-queue?${params.toString()}`
    );
  },

  /**
   * Create a new queue entry
   */
  create: (data: DeploymentQueueCreate) =>
    api.post<DeploymentQueueEntry>("/api/deployment-queue", data),

  /**
   * Get queue entry by ID
   */
  get: (id: string) =>
    api.get<DeploymentQueueEntry>(`/api/deployment-queue/${id}`),

  /**
   * Remove entry from queue
   */
  remove: (id: string) => api.delete(`/api/deployment-queue/${id}`),

  /**
   * Process all pending entries in queue
   */
  process: (data: ProcessQueueRequest) =>
    api.post<ProcessQueueResponse>("/api/deployment-queue/process", data),

  /**
   * Check a deployment CSV without writing anything.
   */
  importPreview: (projectId: string, file: File) =>
    api.upload<CsvImportPreview<DeploymentImportRow>>(
      `/api/deployment-queue/import/preview?project_id=${projectId}`,
      file
    ),

  /**
   * Import a deployment CSV into the queue, all or nothing. The file is sent
   * again rather than the previewed rows, so the backend checks one thing in
   * one way.
   */
  importCsv: (projectId: string, file: File) =>
    api.upload<CsvImportResult>(
      `/api/deployment-queue/import?project_id=${projectId}`,
      file
    ),
};

File: backend/app/api/schemas/deployment_queue.py
"""
Pydantic schemas for Deployment Queue API.

Following DEVELOPERS.md principles:
- Type hints everywhere
- Explicit validation
- Clear separation of create/update/response schemas
"""

from datetime import datetime

from pydantic import BaseModel, Field


class DeploymentQueueBase(BaseModel):
    """Base schema with common deployment queue fields."""

    folder_path: str = Field(..., min_length=1, description="Absolute path to deployment folder")
    site_id: str | None = Field(None, description="Site ID (optional)")
    datetime_offset_seconds: int | None = Field(
        None,
        description="Seconds to add to all file timestamps (null = no adjustment)",
    )
    use_file_mtime_fallback: bool = Field(
        False,
        description=(
            "Fill missing capture dates from each file's modification time. "
            "Never overrides a real capture date"
        ),
    )
    paired_cameras: bool = Field(
        False,
        description=(
            "The subfolders are dependent cameras: cluster "
            "events across them and count effort once"
        ),
    )
    camera_offsets: dict[str, int] = Field(
        default_factory=dict,
        description=(
            "Per-camera seconds added on top of datetime_offset_seconds, "
            "keyed by subfolder name. Paired cameras only"
        ),
    )
    notes: str | None = Field(
        None, max_length=1000, description="Optional deployment notes"
    )
    tags: dict[str, str] = Field(
        default_factory=dict, description="Custom key:value metadata tags"
    )


class DeploymentQueueCreate(DeploymentQueueBase):
    """
    Schema for creating a new queue entry.

    folder_path is required, all other fields are optional.
    """

    project_id: str = Field(..., description="Project ID")
    video_count: int = Field(default=0, description="Number of videos in deployment folder")
    image_count: int = Field(default=0, description="Number of images in deployment folder")


class DeploymentQueueResponse(DeploymentQueueBase):
    """
    Schema for queue entry responses.

    Includes all fields plus generated id and timestamps.
    """

    id: str
    project_id: str
    video_count: int = Field(default=0, description="Number of videos in deployment folder")
    image_count: int = Field(default=0, description="Number of images in deployment folder")
    status: str = Field(..., description="Queue status: pending, processing, completed, failed")
    created_at_utc: datetime
    processed_at_utc: datetime | None = None
    error: str | None = None
    warnings: str | None = Field(
        None,
        description=(
            "Non-fatal warnings from the last ingest, newline-joined. "
            "Populated e.g. when some files were skipped because they "
            "had no extractable capture timestamp."
        ),
    )
    deployment_id: str | None = Field(None, description="Created deployment ID after processing")

    model_config = {"from_attributes": True}  # Enable ORM mode for SQLAlchemy models


class ProcessQueueRequest(BaseModel):
    """
    Schema for processing queue request.

    Specifies which project's queue to process.
    """

    project_id: str = Field(..., description="Project ID to process queue for")
